### PR TITLE
MM-15846 migrating get posts to sync by default

### PR DESCRIPTION
**Pull request by carmo-evan**
#### Summary
migrating get posts to sync by default

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10982

### CODE REVIEW

File: app/post.go [hunk 1]
```diff
@@ -601,19 +601,19 @@ func (a *App) PatchPost(postId string, patch *model.PostPatch) (*model.Post, *mo
...
 	}
-	return result.Data.(*model.PostList), nil
+	return list, nil
 }
 
 func (a *App) GetPosts(channelId string, offset int, limit int) (*model.PostList, *model.AppError) {
-	result := <-a.Srv.Store.Post().GetPosts(channelId, offset, limit, true)
-	if result.Err != nil {
-		return nil, result.Err
+	list, err := a.Srv.Store.Post().GetPosts(channelId, offset, limit, true)
```
Comment (jespino) [inline app/post.go]: Here we can just use `return a.Srv.Store.Post().GetPosts(channelId, offset, limit, true)`

File: store/sqlstore/post_store.go [hunk 2]
```diff
@@ -460,65 +460,61 @@ func (s *SqlPostStore) PermanentDeleteByChannel(channelId string) store.StoreCha
...
+	// Caching only occurs on limits of 30 and 60, the common limits requested by MM clients
+	if allowFromCache && offset == 0 && (limit == 60 || limit == 30) {
+		if cacheItem, ok := s.lastPostsCache.Get(fmt.Sprintf("%s%v", channelId, limit)); ok {
+			if s.metrics != nil {
+				s.metrics.IncrementMemCacheHitCounter("Last Posts Cache")
 			}
+
+			return cacheItem.(*model.PostList), nil
 		} else {
 			if s.metrics != nil {
```
Comment (jespino) [inline store/sqlstore/post_store.go]: Optional: We can simplify a bit here, this block is executed every time the cache misses, and is executed every time we don't use the cache at all, so we can just remove the both else blocks, and move the `if s.metrics....` block right after the end of the `if allowFromCache...` block.

File: store/sqlstore/post_store.go [hunk 3]
```diff
@@ -460,65 +460,61 @@ func (s *SqlPostStore) PermanentDeleteByChannel(channelId string) store.StoreCha
...
-			list := model.NewPostList()
+	var err *model.AppError
+	list := model.NewPostList()
 
-			for _, p := range posts {
-				list.AddPost(p)
-				list.AddOrder(p.Id)
-			}
+	if rpr := <-rpc; rpr.Err != nil {
+		err = rpr.Err
```
Comment (jespino) [inline store/sqlstore/post_store.go]: you can do early return here with `return nil, rpr.Err` and remove the next else. and the same for the next `if` block, leaving the happy path at the least indented level of the function (which is more idiomatic).

File: store/storetest/post_store.go [hunk 4]
```diff
@@ -636,7 +636,7 @@ func testPostStoreGetPostsWithDetails(t *testing.T, ss store.Store) {
 	o5.RootId = o4.Id
 	o5 = (<-ss.Post().Save(o5)).Data.(*model.Post)
 
-	r1 := (<-ss.Post().GetPosts(o1.ChannelId, 0, 4, false)).Data.(*model.PostList)
+	r1, _ := ss.Post().GetPosts(o1.ChannelId, 0, 4, false)
```
Comment (jespino) [inline store/storetest/post_store.go]: We can add here the error checking with `require.Nil(t, err)`

File: store/sqlstore/post_store.go [hunk 5]
```diff
@@ -468,52 +468,48 @@ func (s *SqlPostStore) GetPosts(channelId string, offset int, limit int, allowFr
 	// Caching only occurs on limits of 30 and 60, the common limits requested by MM clients
 	if allowFromCache && offset == 0 && (limit == 60 || limit == 30) {
 		if cacheItem, ok := s.lastPostsCache.Get(fmt.Sprintf("%s%v", channelId, limit)); ok {
-			if s.metrics != nil {
-				s.metrics.IncrementMemCacheHitCounter("Last Posts Cache")
```
Comment (jespino) [inline store/sqlstore/post_store.go]: Here we were counting hits, so you should keep this one. The other two moved out is what I expected :)

File: store/sqlstore/post_store.go [hunk 6]
```diff
@@ -468,52 +468,48 @@ func (s *SqlPostStore) GetPosts(channelId string, offset int, limit int, allowFr
...
 
-	if rpr := <-rpc; rpr.Err != nil {
-		err = rpr.Err
-	} else if cpr := <-cpc; cpr.Err != nil {
-		err = cpr.Err
-	} else {
-		posts := rpr.Data.([]*model.Post)
-		parents := cpr.Data.([]*model.Post)
+	rpr := <-rpc
+	cpr := <-cpc
```
Comment (jespino) [inline store/sqlstore/post_store.go]: This is not the same, not a big difference, but you are waiting for both here, and if the first one fail before, you don't need to wait for the second one, my suggestion would be something like:

```go
rpr := <-rpc
if rpr.Err != nil {
    return nil, rpr.Err
}
cpr := <-cpc
if cpr.Err != nil {
    return nil, cpr.Err
}
```

File: store/storetest/post_store.go [hunk 7]
```diff
@@ -698,13 +699,13 @@ func testPostStoreGetPostsWithDetails(t *testing.T, ss store.Store) {
 	_ = (<-ss.Post().Save(o6)).Data.(*model.Post)
 
 	// Should only be 6 since we hit the cache
-	r3 := (<-ss.Post().GetPosts(o1.ChannelId, 0, 30, true)).Data.(*model.PostList)
+	r3, _ := ss.Post().GetPosts(o1.ChannelId, 0, 30, true)
```
Comment (jespino) [inline store/storetest/post_store.go]: By the way, you can insert the `require.Nil(t, err)` here and in the other similar places, if you want :)

File: app/channel_test.go [hunk 8]
```diff
@@ -452,7 +452,9 @@ func TestAddChannelMemberNoUserRequestor(t *testing.T) {
 	}
 	assert.Equal(t, groupUserIds, channelMemberHistoryUserIds)
 
-	postList := store.Must(th.App.Srv.Store.Post().GetPosts(channel.Id, 0, 1, false)).(*model.PostList)
+	postList, err := th.App.Srv.Store.Post().GetPosts(channel.Id, 0, 1, false)
+	require.Nil(t,err)
```
Comment (gabrieljackson) [inline app/channel_test.go]: A small note that `require.NoError()` will also work for times like this. It reads a little better when viewing the test.

**Comment by gabrieljackson** [comment]
Merged. Thanks @carmo-evan!

**Pull request merged**